### PR TITLE
Do not discard the configuration input listener before the configuration cache entry is written

### PR DESCRIPTION
**Pull request by h0tk3y**
Fixes https://github.com/gradle/gradle/issues/23249

### CODE REVIEW

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt [hunk 1]
```diff
@@ -355,6 +350,10 @@ class DefaultConfigurationCache internal constructor(
             }
         }
 
+        // TODO - fingerprint should be collected until the state file has been written, as user code can run during this process
+        // Moving this is currently broken because the Jar task queries provider values when serializing the manifest file tree and this
+        // can cause the provider value to incorrectly be treated as a task graph input
+        Instrumented.discardListener()
```
Comment (bamboo) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt:356]: ⚙️ This already happens in `doneWithWork` and could be removed from here.

**Review by bamboo — CHANGES_REQUESTED** [review]

**Comment by h0tk3y** [comment]
@bot-gradle test RFR

**Comment by h0tk3y** [comment]
@bot-gradle test RFR

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

**Comment by h0tk3y** [comment]
@bot-gradle cancel

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

**Comment by h0tk3y** [comment]
@bot-gradle test RFR

**Comment by h0tk3y** [comment]
@bot-gradle test RFR

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheInputListenerLifecycleIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+import org.gradle.initialization.StartParameterBuildOptions
+
+class ConfigurationCacheInputListenerLifecycleIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    def "configuration inputs are tracked during task graph serialization"() {
+        given:
+        def configurationCache = newConfigurationCacheFixture()
+
+        and: "a project that has a provider having undeclared configuration inputs, which is linked to a task"
```
Comment (bamboo) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheInputListenerLifecycleIntegrationTest.groovy:28]: 👍

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheInputListenerLifecycleIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            configurationCache.assertStateStored()
+            testDirectory.file("build/out.txt").text == "yes"
+        } else {
+            // This is incorrect behavior of configuration cache that is a result of using the opt-out flag
+            configurationCache.assertStateLoaded()
+            testDirectory.file("build/out.txt").text == "no"
+        }
+
+        where:
+        isOptOut | _
```
Comment (bamboo) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheInputListenerLifecycleIntegrationTest.groovy:78]: 👍

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheInputListenerLifecycleIntegrationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        true     | _
+        false    | _
+    }
+
+    def 'switching the opt-out flag should invalidate the configuration cache entry'() {
```
Comment (bamboo) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheInputListenerLifecycleIntegrationTest.groovy:83]: 👍

File: subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java [hunk 5]
```diff
@@ -528,19 +529,34 @@ public void applyTo(Value value, StartParameterInternal settings, Origin origin)
         }
     }
 
+    public static class ConfigurationCacheIgnoreInputsInTaskGraphSerialization extends BooleanBuildOption<StartParameterInternal> {
+
+        public static final String PROPERTY_NAME = "org.gradle.configuration-cache.inputs.unsafe.ignore-in-serialization";
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java:534]: 👍

**Review by bamboo — APPROVED** [review]
💯 🎉

**Pull request merged**